src/providers/postgres/qgspostgresprovider.cpp

```diff
@@ -1515,23 +1507,37 @@ bool QgsPostgresProvider::hasSufficientPermsAndCapabilities()
       mEnabledCapabilities |= QgsVectorDataProvider::SelectAtId;
     }
 
+    QString sql = QStringLiteral(
+                    "SELECT "
+                    "has_table_privilege(%1,'SELECT')," // 0 (select priv)
+                    "pg_is_in_recovery(),"              // 1 (in recovery)
+                    "current_schema() "                 // 2 (current schema)
+                  ).arg( quotedValue( mQuery ) );
+
+
     if ( connectionRO()->pgVersion() >= 80400 )
     {
-      sql += QString( ",has_any_column_privilege(%1,'UPDATE')" // 5
-                      ",%2" ) // 6
-             .arg( quotedValue( mQuery ),
-                   mGeometryColumn.isNull()
-                   ? QStringLiteral( "'f'" )
-                   : QStringLiteral( "has_column_privilege(%1,%2,'UPDATE')" )
-                   .arg( quotedValue( mQuery ),
-                         quotedValue( mGeometryColumn ) )
-                 );
+      sql += QString(
+               ",has_any_column_privilege(%1,'INSERT')" // 3 (insert priv)
+               ",has_table_privilege(%1,'DELETE')"      // 4 (delete priv)
+               ",has_any_column_privilege(%1,'UPDATE')" // 5 (update priv)
+               ",%2"                                    // 6 (geom upd priv)
+             ).arg( quotedValue( mQuery ),
+                    mGeometryColumn.isNull()
+                    ? QStringLiteral( "'f'" )
+                    : QStringLiteral( "has_column_privilege(%1,%2,'UPDATE')" )
+                    .arg( quotedValue( mQuery ),
+                          quotedValue( mGeometryColumn ) )
+                  );
     }
     else
     {
-      sql += QString( ",has_table_privilege(%1,'UPDATE')" // 5
-                      ",has_table_privilege(%1,'UPDATE')" ) // 6
-             .arg( quotedValue( mQuery ) );
+      sql += QString(
```

Comment:
Maintaining PostgreSQL < 8.4 code paths looks like self inflicted pain at that point, given it has been long unsupported.

Reply:
Agreed. but it's better done in an ad-hoc PR